spec/mem/tree_borrows/tree.md

```diff
@@ -130,5 +164,78 @@ impl Node {
+    /// `path` is the path from `self` to the `node`.
+    fn get_node(&mut self, path: Path) -> Result<Node> {
+        let Some((sub_root_id, sub_path)) = path.split_first() else {
+            return ret(*self);
+        };
+
+        if self.children.len() == Int::ZERO {
+            panic!("Node::get_child_location_states: invalid child path");
+        }
+
+        self.children[sub_root_id].get_node(sub_path)
+    }
+
+}
+```
+
+In addition, we implement some methods for protector-related semantics.
+```rust
+impl Node {
+    /// Release the protector, and perform a special access for the protector end semantics.
+    /// Recusively do state transition on all foreigns of the protected node.
+    /// `path` is the path from `self` to the proctected node.
+    /// `path` is None when the protected node is not a descendant of `self`.
+    /// `location_states` are the location states of the protected node.
+    fn release_protector(
+        &mut self,
+        path: Option<Path>, // self -> protected node
+        location_states: &List<LocationState>,
+        active_calls: &Map<CallId, List<TreeBorrowsProvenance>>,
+    ) -> Result {
+        let node_relation = if path.is_some() { NodeRelation::Child } else { NodeRelation::Foreign };
+
+        // If `self` is the protected node, we are done.
+        if path.is_some_and(|p| p.is_empty()) {
+            self.protected = Protected::No;
+            return ret(());
+        }
+
+        for offset in Int::ZERO..location_states.len() {
+            let LocationState { accessed, permission } = location_states[offset];
+
+            // If the location has never been accessed, there is no need to perform an access here.
+            if accessed != Accessed::Yes { continue; }
+
+            // If there was a write to the location (i.e. if the permission is Active),
+            // we perform a write access here. Otherwise, we perform a read access here.
+            let access_kind = match permission {
+                Permission::Active => AccessKind::Write,
+                _ => AccessKind::Read,
+            };
+
+            self.location_states.mutate_at(Int::from(offset), |location_state|{
+                location_state.transition(access_kind, node_relation, self.protected)
+            })?;
+        }
+
+        for child_id in Int::ZERO..self.children.len() {
+            let sub_path = match path.and_then(|p| p.split_first()) {
+                Some((head, tail)) if head == child_id => Some(tail),
+                _ => None,
+            };
+
+            self.children.mutate_at(child_id, |child| { child.release_protector(sub_path, &location_states, active_calls) })?;
+        }
+
+        ret(())
+    }
+
+    /// Recusively check whether there is a protected node in `self` and all its descendants.
+    fn contain_strong_protector(&self) -> bool {
+        self.protected != Protected::No || self.children.any(|child| child.contain_strong_protector())
+    }
```

Comment:
What does this method do? It should check for `self.protected == Protected::Strong`, since it's used during deallocation to check for strong protectors, shouldn't it?

Also the comment is wrong.

Reply:
oops, thanks for pointing out this :)

We need a test like `deallocate_weak_protector.rs`, but unfortunately we cannot translate `Box` to minirust.
Maybe a `FIXME` for adding this test later.